src/metrics/task_metrics_rewriter.rs

```diff
@@ -1,82 +1,235 @@
+use crate::metrics::MetricsCollectorResult;
+use crate::metrics::TaskMetricsCollector;
 use crate::metrics::proto::MetricsSetProto;
 use crate::metrics::proto::metrics_set_proto_to_df;
+use crate::protobuf::StageKey;
+use crate::stage::DisplayCtx;
+use crate::stage::Stage;
+use bytes::Bytes;
+use datafusion::common::HashMap;
 use datafusion::common::tree_node::Transformed;
 use datafusion::common::tree_node::TreeNode;
 use datafusion::common::tree_node::TreeNodeRecursion;
 use datafusion::common::tree_node::TreeNodeRewriter;
 use datafusion::error::Result;
 use datafusion::physical_plan::ExecutionPlan;
 use datafusion::physical_plan::internal_err;
+use datafusion::physical_plan::metrics::MetricsSet;
 use std::sync::Arc;
+use std::vec;
 
-/// TaskMetricsRewriter is used to enrich a task with metrics by re-writing the plan using [MetricsWrapperExec] nodes.
+/// Rewrites a distributed plan with metrics. Does nothing if the root node is not a [DistributedExec].
+/// Returns an error if the distributed plan was not executed.
+pub fn rewrite_distributed_plan_with_metrics(
+    plan: Arc<dyn ExecutionPlan>,
+) -> Result<Arc<dyn ExecutionPlan>> {
+    if !plan.as_any().is::<DistributedExec>() {
+        return Ok(plan);
+    }
+
+    // Metrics are available in the prepared plan of the DistributedExec.
+    let executed_plan = plan
+        .as_any()
+        .downcast_ref::<DistributedExec>()
+        .unwrap()
+        .pepared_plan()?;
+
+    // Collect metrics from the DistributedExec.
+    // TODO: Should we move this into the DistributedExec itself or a new ExplainAnalyzeExec?
+    let MetricsCollectorResult {
+        task_metrics,       // Metrics for the DistributedExec plan
+        input_task_metrics, // Metrics for all child stages / tasks.
+    } = TaskMetricsCollector::new().collect(executed_plan.clone())?;
+
+    let metrics_collection = Arc::new(input_task_metrics);
+
+    let transformed = plan.transform_down(|plan| {
+        // Transform all stages using NetworkShuffleExec and NetworkCoalesceExec as barriers.
+        if let Some(network_boundary) = plan.as_network_boundary() {
+            return match network_boundary.input_stage() {
+                Some(stage) => {
+                    // TODO: This transform is a bit inefficient because we traverse the plan nodes more than once.
+                    // For now, this is acceptable for clarity.
+                    let plan_with_metrics =
+                        StageMetricsRewriter::new(stage, metrics_collection.clone())?.rewrite()?;
+                    Ok(Transformed::new(
+                        network_boundary.with_input_stage(Stage::new(
+                            stage.query_id,
+                            stage.num,
+                            plan_with_metrics,
+                            stage.tasks.len(),
+                        ))?,
+                        true,
+                        TreeNodeRecursion::Continue,
+                    ))
```

Comment:
You can use the `Transformed::yes` shorthand for this:
```suggestion
                    Ok(Transformed::yes(network_boundary.with_input_stage(
                        Stage::new(
                            stage.query_id,
                            stage.num,
                            plan_with_metrics,
                            stage.tasks.len(),
                        ),
                    )?))
```